Fix elasticnect cv sample weight

snath-xoc opened the pull request:
Fixes #22914

What does this implement/fix? Explain your changes.

Adapted from the pull request #23045 by s-banach and #29308 from snath-xoc

Modifies _alpha_grid function in linear_model._coordinate_descent to accept a sample_weight argument and implements changes to be compatible with _preprocess_data

TODO
- [x] Check why test_enet_cv_sample_weight_correctness fails for 6-csc_matrix-True and 6-csc_array-True

s-banach committed:
Update _alpha_grid to take sample_weight

It seems like this single call to _preprocess_data suffices in all cases.

s-banach committed:
Add a simple test for alpha_max with sample_weight

This tiny example was given in https://github.com/scikit-learn/scikit-learn/issues/22914.
The test merely asserts that alpha_max is large enough to force the coefficient to 0.

s-banach committed:
Update test

As per reviewer's suggestions:
(1) Clarify eps=1.
(2) Parameterize `fit_intercept`.

s-banach committed:
Clarify _alpha_grid.

(1) Give the name `n_samples` to the quantity `X.shape[0]`.
(2) Clarify that `y_offset` and `X_scale` are not used, since these are already applied to the data by `_preprocess_data`.

s-banach committed:
Clarify notation

s-banach committed:
Use Xy if it is provided.

s-banach committed:
Update test, check alpha_max is not too large

s-banach committed:
Fix test that alpha_max is not too large.

s-banach committed:
Test alpha_max without sample_weight.

snath-xoc committed:
fix elasticnetcv sample weighting adapted from previous commit by s-banach

snath-xoc committed:
Update _preprocess_data inputs in _coordinate_descent.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

snath-xoc committed:
added tests for repeated vs weighted on cyclic ElasticNetCV and modified alpha_grid_ to accommodate for MultitaskCV y shape

ogrisel committed:
Merge branch 'main' into fix_elasticnet_cv_sample_weight

snath-xoc committed:
added to changelog and changed seeding in tests

snath-xoc committed:
[all random seeds] test_enet_cv_sample_weight

ogrisel committed:
Merge branch 'main' into fix_elasticnet_cv_sample_weight

ogrisel committed:
Revert unrelated changes

snath-xoc committed:
merged test into test_enet_cv_sample_weight_correctness

snath-xoc committed:
changed sample weight to be explicitly set as integers in sklearn/linear_model/tests/test_coordinate_descent/test_enet_cv_sample_weight_correctness

Commit:
add sample weights into default score calculation under _log_reg_scoring_path

snath-xoc committed:
[all random seeds] test_enet_cv_sample_weight

snath-xoc committed:
[all random seeds] test_enet_cv_sample_weight

snath-xoc committed:
Merge branch 'main' into fix_elasticnet_cv_sample_weight

snath-xoc committed:
modified test_coordinate_descent for X, y, group names

snath-xoc committed:
restoring some files

snath-xoc committed:
add global_random_seed

ogrisel committed:
Merge branch 'main' into fix_elasticnect_cv_sample_weight

ogrisel commented:
I merged the current `main` to be able to see an up to date output of the CI during my review.

Please to do not forget to pull the merge commit before resuming work on this PR.

Also, I closed #29308 since this PR (#29442) has taken most of the comment of the past review I did in #29308. Next time, there is no need to create a new branch and a new PR. Instead address the review comments by pushing commits to the same branch to get the PR updated.

It would make the review process easier to follow by having a single linear discussion.

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1304,55 +1305,60 @@ def test_enet_sample_weight_consistency(
-    assert reg_sw.intercept_ == pytest.approx(reg.intercept_)
+    assert_allclose(reg_with_weights.mse_path_, reg_with_repetitions.mse_path_)
+    assert_allclose(reg_with_weights.alphas_, reg_with_repetitions.alphas_)
+    assert reg_with_weights.alpha_ == reg_with_repetitions.alpha_
```
Since `alpha_` is a floating point scalar, we don't want this assertion to fail due to rounding errors.

For arrays with floating point values we use `assert_allclose` instead of `assert_array_equal` to tolerate rounding errors. For assertion on scalar values, we can either use `assert_allclose` but we can more explicitly use:

```suggestion
    assert reg_with_weights.alpha_ == pytest.approx(reg_with_repetitions.alpha_)
```

This makes it explicit that we expect `alpha_` to be a scalar instead of an array.

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1390,6 +1396,26 @@ def test_enet_cv_grid_search(sample_weight):
     assert reg.alpha_ == pytest.approx(gs.best_params_["alpha"])
 
 
+@pytest.mark.parametrize("sparseX", [False, True])
```
I would rename this to `X_is_sparse` or `sparse_input_data`.

ogrisel reviewed (commented):
I have the feeling that this PR is almost ready to go contrary to what the description and the draft state imply. Could you please update the TODO section of the description of the PR to reflect what actually remains to do with the current state of the PR (if anything is actually left to do)?

If there is nothing left to do, please hit the "ready to review" button to attract other maintainers attention, otherwise you might never get reviews.

The CI test failure seems to be caused by a missing `pytest.approx` as explained below.

snath-xoc committed:
updated comments and changelog

snath-xoc committed:
[all random seeds]

ogrisel commented:
Some comments about the last commit 69646d9 and its CI results:

- the `[all random seeds]` commit message flag is only valid if it is followed by specific test names as explained in the pinned issue (#28959).
- you can and should run those variable seed tests locally first with:
 
```
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -vlx -k test_enet_cv_sample_weight_correctness
```

this will allow to detect brittle choice of tolerance level without having to wait for the CI to run and iterate faster.

- in particular a test in the macOS `pylatest_conda_mkl_no_openmp` build has failed because:
   - elasticnet / lasso use an l1 penalty, so it's possible to have exact zeros or near zero values in `coef_`
   - `assert_allclose` only specifies a relative tolerance by default (`rtol`) and the defaullt value for the absolute tolerance level (`atol`) is null;
   - as a result comparing a very small near zero value to another near zero value with a difference caused by machine precision level rounding errors can lead to a very large relative error that exceeds the default value of `rtol`.
   - to avoid this, you can set `atol` to a smallish value yet a few orders of magnitude larger than machine precision, e.g. `atol=1e-10` or even `atol=1e-8` if this test still fails with


The "circle ci doc" reports the following two new warnings with the changelog entry and as a result this CI is red and needs to be fixed:

```
/home/circleci/project/doc/whats_new/v1.6.rst:139: WARNING: Title underline too short.
/home/circleci/project/doc/whats_new/v1.6.rst:141: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

ogrisel commented:
Also note: when I put suggestions in the review comments, you can accept/apply them via the github interface directly (if you agree): github will add the commits to your branch in your fork. You can then pull those commits by to your local branch to further iterate on the PR locally.

snath-xoc committed:
[all random seeds]
test_enet_cv_sample_weight_correctness

snath-xoc committed:
trigger regular CI

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -264,7 +265,9 @@ def test_lasso_cv():
     # check that they also give a similar MSE
     mse_lars = interpolate.interp1d(lars.cv_alphas_, lars.mse_path_.T)
-    assert_allclose(mse_lars(clf.alphas_[5]).mean(), clf.mse_path_[5].mean(), rtol=1e-2)
+    np.testing.assert_approx_equal(
```
We rather use `assert_allclose` and set `rtol` and `atol` instead of specifying the tolerance using a number of significant decimals.

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1444,9 +1465,29 @@ def test_enet_cv_sample_weight_consistency(
+    # Test alpha_max makes coefs zero.
+    reg = ElasticNetCV(n_alphas=1, cv=2, eps=1, fit_intercept=fit_intercept)
+    reg.fit(X, y, sample_weight=sample_weight)
+    assert_almost_equal(reg.coef_, 0)
```
Here again, let's use `assert_allclose` for the sake of consistency.

snath-xoc committed:
Merge branch 'main' into fix_elasticnect_cv_sample_weight

snath-xoc committed:
[all random seeds]
test_lasso_cv
test_enet_alpha_max_sample_weight
test_enet_cv_sample_weight_correctness

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1304,55 +1305,73 @@ def test_enet_sample_weight_consistency(
+
+    We fit the same model twice, once with weighted training data, once with repeated
+    data points in the training data and check that both models converge to the
+    same solution.
```
Text formatting nitpick.

```suggestion
    same solution.

```

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1304,55 +1305,73 @@ def test_enet_sample_weight_consistency(
+    reg_with_repetitions.fit(X_with_repetitions, y_with_repetitions)
 
     # ensure that we chose meaningful alphas, i.e. not boundaries
-    assert alphas[0] < reg.alpha_ < alphas[-1]
```
This assertion should be kept (but adapted to `reg_with_repetitions.alpha_`) and the comment ("Ensure that we chose meaningful alphas, i.e. not boundaries.") should be kept to this particular assertion.

ogrisel reviewed (approved):
A final pass of feedback. Once addressed, LGTM for merge.

ogrisel commented:
Similarly to `LogisticRegressionCV`, `LinearModelCV` has the following estimator tag:

```python
    def _more_tags(self):
        # Note: check_sample_weights_invariance(kind='ones') should work, but
        # currently we can only mark a whole test as xfail.
        return {
            "_xfail_checks": {
                "check_sample_weights_invariance": (
                    "zero sample_weight is not equivalent to removing samples"
                ),
            }
        }
```

but this cannot be removed without changing `check_sample_weights_invariance` as explained in:

https://github.com/scikit-learn/scikit-learn/pull/29419#issuecomment-2331025642

snath-xoc committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel committed:
Fix linting problem

ogrisel commented:
@snath-xoc I pushed a quick fix for the linting problem that resulted from the merge of one of my github-edited suggestion.

The following comment still need be addressed before final review / merge:

https://github.com/scikit-learn/scikit-learn/pull/29442#discussion_r1743917261

snath-xoc committed:
updated changelog

ogrisel commented on doc/whats_new/v1.6.rst:
```diff
@@ -229,6 +229,11 @@ Changelog
+  take sample weights into accounts to define the search grid for the internally tuned
+  `alpha` hyper-parameter. :pr:`29442` by :user:`John Hopfensperger <s-banach> and
+  :user:`Shruti Nath <snath-xoc>`.
+
```
I think fixes are ranked after API and new features. Look at the instructions at above in this file for details.

snath-xoc committed:
moved fix in changelog to after API

ogrisel committed:
Restore non-degenerate alpha check in test

ogrisel commented:
@snath-xoc for information I pushed 2b99c9b2824896b6e0989aed250aa2c61e2d5ca1 to address the remaining unaddressed review comment (https://github.com/scikit-learn/scikit-learn/pull/29442#discussion_r1743917261).

This PR LGTM.

jeremiedbb commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1304,55 +1305,84 @@ def test_enet_sample_weight_consistency(
+    assert_allclose(reg_with_weights.alphas_, reg_with_repetitions.alphas_)
+    assert reg_with_weights.alpha_ == pytest.approx(reg_with_repetitions.alpha_)
+
+    # Ensure that the models selected a non-boundary alpha to make sure that
+    # the sample_weight equivalence test is non-degenerate.
+    alphas = reg_with_weights.alphas_
+    alpha = reg_with_weights.alpha_
+    assert alphas[0] > alpha > alphas[-1]
```
Some tests are failing because the selected alpha is actually a boundary.

ogrisel replied:
Yes, I tried to tweak the dataset generation process but I could not find a way to make it robustly select a non-boundary alpha for all the admissible seeds (especially with the default value of the `eps=1e-3` parameter of the alpha grid generator.

I manually check that it is not boundary most of the time though. So the test is not degenerate.

ogrisel committed:
Remove alpha boundary check

jeremiedbb reviewed (approved):
LGTM, thanks !
